training/coach.py

```diff
@@ -250,18 +308,55 @@ def get_normed_embeds(self, text_prompts: List[str]) -> torch.Tensor:
-            live_negative = self.query_vlm(sampled_image)
-            self.cfg.negative_classes.append(live_negative)
-        elif self.cfg.gradual_negatives:
-            random.shuffle(self.cfg.negative_classes)
-            self.cfg.negative_pool = copy(self.cfg.negative_classes)
-            self.cfg.negative_classes = [self.cfg.negative_pool.pop(0)]
+
+        # to avoid generating image for specific class too many times
+        count = 0
+        image_embs = None
+        if self.cfg.image_feature and not self.cfg.live_negatives:
+            raise RuntimeError("image feature must with live_negatives")
+        elif self.cfg.image_feature:
+            # guarantee image_embs will have corresponding feature with negative_classes
+            image_embs = defaultdict(list)
+            negative_classes = defaultdict(list)
+
+        # TODO: move negative related image and feature to method
+        if not self.cfg.image_feature:
+            sampled_images = self.save_images(save_dir=self.cfg.images_root, save_prefix=f'init_images')
+            if self.cfg.live_negatives and len(self.cfg.negative_classes) == 0:
+                live_negatives = self.query_vlm(sampled_images)
+                self.cfg.negative_classes.append(live_negative)
+            elif self.cfg.gradual_negatives:
+                random.shuffle(self.cfg.negative_classes)
+                self.cfg.negative_pool = copy(self.cfg.negative_classes)
+                self.cfg.negative_classes = [self.cfg.negative_pool.pop(0)]
+        else:
+            from training.templates import PREFIXES
+            for temp in ConceptDataset(
+                placeholder_token=self.cfg.placeholder_token,
+                learnable_property=self.cfg.learnable_property
+            ).templates:
+                templates = [temp] if '{a}' not in temp else []
+                if '{a}' in temp:
+                    for a in PREFIXES: # a the an
+                        templates.append(temp.format(a=a, token='{token}'))
+                for template in templates:
+                    while len(image_embs[template]) == 0:
+                        temp = []
+                        classes_temp = []
+                        sampled_images = self.save_images(save_dir=self.cfg.images_root, save_prefix=f'init_images', image_embs=temp, template=[template], fix_seed=False)
+                        live_negatives = self.query_vlm(sampled_images)
+                        for live_negative in live_negatives:
+                            if not self.cfg.specific_negatives or any( cluster in live_negative for cluster in self.cfg.specific_negatives):
+                                classes_temp.append(live_negative)
+                            elif image_embs is not None:
+                                del temp[len(classes_temp)]
+
+                        image_embs[template].extend(temp)
+                        negative_classes[template].extend(classes_temp)
```

Comment:
collect image for negative prompt first
1. we will generate image from "mypet" or "pet" prompt
2. put them image feature into image_embs for corresponding class like "a photo of a {token}" but not "a photo of a pet" or "a photo of a dog"
3. TODO: put them into training process instead of collecting them before training